### PR TITLE
Type fixes

### PR DESCRIPTION
**Pull request by segiddins**

### CODE REVIEW

File: lib/literal/data_structure.rb [hunk 1]
```diff
@@ -11,13 +11,29 @@ def self.from_pack(payload)
 	end
 
 	def [](key)
-		instance_variable_get(:"@#{key}")
+		case key
+		when Symbol
+		when String
+			key = key.intern
```
Comment (stevegeek) [inline lib/literal/data_structure.rb:17]: Im personally not a fan of 'indifferent access' on structures etc. Could this be opt-in maybe? Or another specific data structure base class ala "data structure with indifferent access"?

Comment (segiddins) [reply lib/literal/data_structure.rb:17]: I was mirroring how the stdlib Struct/Data work

Comment (joeldrapper) [reply lib/literal/data_structure.rb:17]: I don’t feel strongly about this one way or the other. I wouldn’t add indifferent access to Literal::Hash, but I think it’s okay here.

Comment (stevegeek) [reply lib/literal/data_structure.rb:17]: For Struct I can see the appeal of mirroring stdlib. Data however doesn't implement [] if I'm not mistaken, maybe we could just add this to Struct?

**Comment by segiddins** [comment]
stdlib:

```ruby
#!/usr/bin/env ruby

[[Struct, :new], [Data, :define]].each do |sc, m|
  Person = sc.send(m, :name, :age)
  
  person_ = Person.new("abc", 123)
  
  %w[
    .name
    .age
    [:name]
    [:age]
    ["name"]
    ["age"]
    [0]
    [3]
    [-1]
    .foo
    .foo=1
    [:foo]
    [[]]
    [:foo]=1
    .to_a
    .to_h
    .deconstruct
    .deconstruct_keys([])
    .deconstruct_keys(nil)
    .deconstruct_keys(["name"])
  ].each do |b|
    person = person_.dup
    print("person#{b} => ")
    result = begin
      binding.eval("person#{b}")
    rescue Exception => e
      e
    end
    print result.inspect
    puts " # #{person.inspect}"
  end
end
```

```
person.name => "abc" # #<struct Person name="abc", age=123>
person.age => 123 # #<struct Person name="abc", age=123>
person[:name] => "abc" # #<struct Person name="abc", age=123>
person[:age] => 123 # #<struct Person name="abc", age=123>
person["name"] => "abc" # #<struct Person name="abc", age=123>
person["age"] => 123 # #<struct Person name="abc", age=123>
person[0] => "abc" # #<struct Person name="abc", age=123>
person[3] => #<IndexError: offset 3 too large for struct(size:2)> # #<struct Person name="abc", age=123>
person[-1] => 123 # #<struct Person name="abc", age=123>
person.foo => #<NoMethodError: undefined method `foo' for an instance of Person> # #<struct Person name="abc", age=123>
person.foo=1 => #<NoMethodError: undefined method `foo=' for an instance of Person> # #<struct Person name="abc", age=123>
person[:foo] => #<NameError: no member 'foo' in struct> # #<struct Person name="abc", age=123>
person[[]] => #<TypeError: no implicit conversion of Array into Integer> # #<struct Person name="abc", age=123>
person[:foo]=1 => #<NameError: no member 'foo' in struct> # #<struct Person name="abc", age=123>
person.to_a => ["abc", 123] # #<struct Person name="abc", age=123>
person.to_h => {:name=>"abc", :age=>123} # #<struct Person name="abc", age=123>
person.deconstruct => ["abc", 123] # #<struct Person name="abc", age=123>
person.deconstruct_keys([]) => {} # #<struct Person name="abc", age=123>
person.deconstruct_keys(nil) => {:name=>"abc", :age=>123} # #<struct Person name="abc", age=123>
person.deconstruct_keys(["name"]) => {"name"=>"abc"} # #<struct Person name="abc", age=123>
Untitled 7.rb:4 <coderunner://Untitled%207.rb#(71,0)!>: warning: already initialized constant Person
Untitled 7.rb:4 <coderunner://Untitled%207.rb#(71,0)!>: warning: previous definition of Person was here
person.name => "abc" # #<data Person name="abc", age=123>
person.age => 123 # #<data Person name="abc", age=123>
person[:name] => #<NoMethodError: undefined method `[]' for an instance of Person> # #<data Person name="abc", age=123>
person[:age] => #<NoMethodError: undefined method `[]' for an instance of Person> # #<data Person name="abc", age=123>
person["name"] => #<NoMethodError: undefined method `[]' for an instance of Person> # #<data Person name="abc", age=123>
person["age"] => #<NoMethodError: undefined method `[]' for an instance of Person> # #<data Person name="abc", age=123>
person[0] => #<NoMethodError: undefined method `[]' for an instance of Person> # #<data Person name="abc", age=123>
person[3] => #<NoMethodError: undefined method `[]' for an instance of Person> # #<data Person name="abc", age=123>
person[-1] => #<NoMethodError: undefined method `[]' for an instance of Person> # #<data Person name="abc", age=123>
person.foo => #<NoMethodError: undefined method `foo' for an instance of Person> # #<data Person name="abc", age=123>
person.foo=1 => #<NoMethodError: undefined method `foo=' for an instance of Person> # #<data Person name="abc", age=123>
person[:foo] => #<NoMethodError: undefined method `[]' for an instance of Person> # #<data Person name="abc", age=123>
person[[]] => #<NoMethodError: undefined method `[]' for an instance of Person> # #<data Person name="abc", age=123>
person[:foo]=1 => #<NoMethodError: undefined method `[]=' for an instance of Person> # #<data Person name="abc", age=123>
person.to_a => #<NoMethodError: undefined method `to_a' for an instance of Person> # #<data Person name="abc", age=123>
person.to_h => {:name=>"abc", :age=>123} # #<data Person name="abc", age=123>
person.deconstruct => ["abc", 123] # #<data Person name="abc", age=123>
person.deconstruct_keys([]) => {} # #<data Person name="abc", age=123>
person.deconstruct_keys(nil) => {:name=>"abc", :age=>123} # #<data Person name="abc", age=123>
person.deconstruct_keys(["name"]) => {"name"=>"abc"} # #<data Person name="abc", age=123>
```

**Comment by joeldrapper** [comment]
I’m merging this as it is, but I think we need to consider the change to `Literal::Data#[]` carefully before cutting a release and may need to undo it. Removing `#[]` is a breaking change which would require 2.0.

**Pull request merged**